Adding Canonicalizer component of pipeline

m-cheung opened the pull request:
This PR depends on #69

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+  name: 'Canonicalizer',
+
+  documentation: `This component expects to receive AST nodes from various
+    sources. Canonical versions of interface definitions will be created
```
Better description of a "source"? I think this means from different web platform engines, or IDL fragment repositories.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+
+  documentation: `This component expects to receive AST nodes from various
+    sources. Canonical versions of interface definitions will be created
+    from AST nodes of the same source.
```
nit: Wrap lines when you have to. If you want a paragraph break, include an empty line.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+  documentation: `This component expects to receive AST nodes from various
+    sources. Canonical versions of interface definitions will be created
+    from AST nodes of the same source.
+    Please note that this step cannot be distributed. All computations
```
More precisely, canonicalization for the same IDL fragment repository cannot be distributed. Distributing between repositories would be a perfectly reasonable thing to do.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+  properties: [
+    {
+      class: 'Function',
+      documentation: `Function is called after waitTime seconds has passed
```
Please explain the **why** here. I think it has to do with being unsure when the canonicalizer is finished receiving fragments and can "finalize" its database of canonical fragments.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+    {
+      documentation: `A map of maps containing canonical IDL files. The
+        key of the map corresponds to the renderer (source) from which the
+        files were extract. For example, rendererMap['blink'] is a map that
```
nit: `rendererMap_`

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+  ],
+
+  properties: [
+    {
```
Please add a `TODO`. This component should not be responsible for observing the pipeline; eventually `onDone` and `waitTime` should be removed or replace with a subscription to a component that *is* responsible for observing the pipeline.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+      name: 'rendererMap_',
+      factory: function() { return {}; },
+    },
+    {
```
Please move this up to after `waitTime`. That will keep together the state related to observing the pipeline.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+      name: 'addFragment',
+      code: function(source, ast, metadata) {
+        // Determine if Map exists for this source. Create one if non-existent.
+        var map = this.rendererMap_[source];
```
`var map = this.rendererMap_[source] = this.rendererMap_[source] || {};`

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        var def = ast.definition;
+        // Ignore definition if Enum or Typedef.
```
Is this the right thing to do? Should this component store them? Warn or throw on name reuse? Leaving a `TODO` might be sufficient here, but I'm not convinced that dropping these is the right choice.

m-cheung replied:
I thought we previously discussed that `Enum` and `Typedef` was not required. If they are, I can add them into the map and add checks to ensure that we don't merge if they are found during lookup.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+
+        // Find any existing interface with same id (name).
+        var clone = ast.clone();
+        var result = map[ast.id];
```
Let's add an assertion that `ast.id` is not falsey.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+        if (!result) {
+          // No interface with the same name was found.
+          // We add a copy of ourselves into the map.
+          clone.ref = [metadata];
```
`ref` isn't defined in the classes being used here, is it? Either it should be added, or this component should be storing a wrapper object that contains something like `{references, ast}`.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+
+          // Determine if we have multiple non-partial fragments.
+          if (!isCanonicalPartial && !isDefPartial) {
+            console.error(`Two non-partial fragments!
```
Please include the class name in the message, and use `imports: [ 'error' ] ... this.error(...)`.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+          // Determine if we have multiple non-partial fragments.
+          if (!isCanonicalPartial && !isDefPartial) {
+            console.error(`Two non-partial fragments!
+              In Map: ${result.ref},
```
If we want this to be printed nicely in the debugger, we probably want to call `foam.json.objectify()`. If we want it printed nicely in the console, we probably want `foam.json.Pretty.stringify()`.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+          if (!isCanonicalPartial && !isDefPartial) {
+            console.error(`Two non-partial fragments!
+              In Map: ${result.ref},
+              Current: ${metadata}`);
```
ditto

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+            console.error(`Two non-partial fragments!
+              In Map: ${result.ref},
+              Current: ${metadata}`);
+            throw new Error('Two non-partial fragments were found for the same interface!');
```
Let's add the class name (of the component) and the entity name (that has the name conflict) in this message.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+            throw new Error('Two non-partial fragments were found for the same interface!');
+          }
+
+          // Copy attributes and members from definition into canonical AST.
```
Did we decide against storing pre-sub-fragment source information? I wonder whether we should store such information. For example, if there are two members with the same name but different <something>, how hard is it to figure out where each came from? Even though this is theoretically computable from other data, adding it here might be worthwhile: Consider a UI component that wants to add linking information; should it have to issue queries against the non-canonical database and reason about the results to discover where a member or extended attr came from?

m-cheung replied:
To be clear, you would like metadata on where each member and attribute came from? It shouldn't be too hard to do, but it may make the code uglier (at least I think it will)

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+          result.ref.push(metadata);
+
+          // Copy over inheritance information.
+          if (!result.definition.inheritsFrom)
```
Feels like error checking might be missing here. What if multiple entities are tagged with a super class? I think partials cannot define one (so, in theory, we should be okay), but we should probably be retaining all the information if we find multiple non-partials.

m-cheung replied:
This assumes that we will process multiple non-partials correct? As I mentioned in another comment, I don't think multiple non-partial fragments are defined (assuming we have a correct set of data).

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+
+          // Update AST partial status if required.
+          if (!isDefPartial) {
+            result.definition.isPartial = false;
```
This seems reasonable, but I'm a little uneasy about the semantics of `isPartial` changing. Rather than meaning "this concrete AST fragment is partial", it means "all of the concrete AST fragments merged into this one are partial".

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+        this.timerMap_[source] = setTimeout(function() {
+          var map = this.rendererMap_[source];
+          var results = [];
+          for (var key in map) {
```
I think this is missing a `map.hasOwnProperty(key)` guard.

mdittmer commented on lib/org/chromium/webidl/CanonicalizerRunner.js:
```diff
@@ -0,0 +1,60 @@
+  requires: ['org.chromium.webidl.Canonicalizer'],
+
+  constants: {
+    MISSING_PROPS: 'Missing required properties in PipelineMessage',
```
ditto from other PRs: Elaborate message with specifics of runner class and what's missing. No need  for a constant at this time.

mdittmer commented on lib/org/chromium/webidl/CanonicalizerRunner.js:
```diff
@@ -0,0 +1,60 @@
+      factory: function() {
+        return this.Canonicalizer.create({
+          onDone: this.output.bind(this),
+          waitTime: this.waitTime,
```
Could `waitTime` ever be changed? If not, this is probably fine. If so, we need a binding.

m-cheung replied:
I did not plan on having `waitTime` changed once it was set. I don't think there's a reason one would need to do that.

mdittmer commented on lib/org/chromium/webidl/CanonicalizerRunner.js:
```diff
@@ -0,0 +1,60 @@
+
+  methods: [
+    function run(message) {
+      // If SUPER threw an error, we are done.
```
May need update as per conversation in previous PRs.

mdittmer commented on lib/org/chromium/webidl/CanonicalizerRunner.js:
```diff
@@ -0,0 +1,60 @@
+
+      // Verify expected parameters are present.
+      if (!asts || !file || !renderer) {
+        this.error(this.fmtErrorMsg(this.MISSING_PROPS));
```
We should report on what's missing, even if it's just by printing out all three.

mdittmer commented on lib/org/chromium/webidl/CanonicalizerRunner.js:
```diff
@@ -0,0 +1,60 @@
+
+      // Insert all of the given ASTs.
+      asts.forEach(function(ast) {
+        var id = file.id.concat('/');
```
Why `concat()` on a String? Is the intention here `file.id + '/'`?

i.e.,

```
`${file.id}/`
```

m-cheung replied:
`file.id` is not a string. It's an array of string.
For files coming from a repository, the id becomes `<repository>/<revision>/<filename>` I believe, and for files coming from the web, it becomes a URL. (I will have to double check this to be sure)

The ID can be change as needed. This is used as metadata (allowing us to determine which sources is the canonical file is made of)

mdittmer commented on test/any/Canonicalizer-test.js:
```diff
@@ -0,0 +1,181 @@
+  });
+
+  it('should return same number of files since we have different interfaces', function(done) {
+    // Setting up files for canonicalization.
```
Consider writing a helper that simply receives a string containing IDL with exactly one top-level entity and returns the unwrapped parse value (`.value[0]`).

mdittmer commented on test/any/Canonicalizer-test.js:
```diff
@@ -0,0 +1,181 @@
+    canonicalizer.addFragment('Gecko', ast, 'First file');
+  });
+
+  it('should throw error if two non-partial interfaces were given for same source', function() {
```
Does this ever happen in practice? We may want to tolerate this issue and emit a warning.

mdittmer replied:
@foolip WDYT about this? Is parsing two non-paritals with the same name an unrecoverable error? Are there cases where we may genuinely wish to consume two current specs that both contain a non-partial definition of the same thing?

m-cheung replied:
So far, it seems like most of the cases where two non-partial interfaces are defined, are when they define the exact same members. Currently, repositories have no occurrences of this. It only happens when we canonicalize WebSpec IDL. Thus, I am inclined to believe that it should not normally happen in practice (if we have a good set of data)

mdittmer replied:
As per offline discussion, I think we can throw with this for now.

mdittmer commented on test/any/CanonicalizerRunner-test.js:
```diff
@@ -0,0 +1,139 @@
+    ResultBox = foam.lookup('org.chromium.webidl.Test.ResultBox');
+  });
+
+  it('should send an error if invalid arguments are received as message', function() {
```
This test pattern crops up with every `PipelineRunner` implementation. Should we write a `-helper.js` function that can be used to capture "wrong obj" and "missing input data" cases?

mdittmer commented on test/any/CanonicalizerRunner-test.js:
```diff
@@ -0,0 +1,139 @@
+
+    runner.run(firstMessage);
+    runner.run(secondMessage);
+    // Expect results to take ~3 seconds to arrive.
```
Add a `TODO` to observe output and/or error boxes here.

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+      name: 'waitTime',
+    },
+    {
+      documentation: `A map of maps containing canonical IDL files. The
```
I don't know if this is a good structure. It feels like you could have one instance of this for each source instead of having one instance that holds a map of things for each source.

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+    },
+    {
+      documentation: `A map of Timers that is used to determine whether waitTime
+        has passed since the last call to addFragment.`,
```
last call for that source?

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+
+        var def = ast.definition;
+        // Ignore definition if Enum or Typedef.
+        if (this.Enum.isInstance(def) || this.Typedef.isInstance(def)) return;
```
It seems bad for the timer to not be reset here.

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+          // Copy attributes and members from definition into canonical AST.
+          result.attrs = result.attrs.concat(clone.attrs);
+          result.definition.members = result.definition.members
+            .concat(clone.definition.members);
```
nit: +4 indent here, not +2

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+            throw new Error('Two non-partial fragments were found for the same interface!');
+          }
+
+          // Copy attributes and members from definition into canonical AST.
```
I think the merge operation should be in a separate method. That will make it easier to test.

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,129 @@
+    {
+      class: 'Function',
+      documentation: `Function is called after waitTime seconds has passed
+        since the last call to addFragment.`,
```
Mention that it's called for each source

m-cheung commented:
@mdittmer PTAL at this when you have time. I believe I have addressed all of the concerns. The tests are currently failing since this PR depends on #69.

Additionally, an instance of `Canonicalizer` is instantiated for each repository / source now. After considering comments from you and @arobins, I think this step should be more clear and cleaner now. Pipeline setup has changed a little due to this as well, and those changes can be seen in the `runner` branch in `main/runner.js`

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,145 @@
+
+  properties: [
+    // TODO: This component should not be responsible for observing the
+    // pipeline. onDone and waitTime should be removed in the futute and
```
nit: typo in future

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,145 @@
+      name: 'waitTime',
+    },
+    {
+      documentation: `A timers that is used to determine whether waitTime
```
nit: a timer

arobins commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,145 @@
+    {
+      documentation: `Takes a given AST and extracts the definition members
+        from the AST. The members are then placed into the map corresponding to
+        the given rendering engine and definition.`,
```
nit: remove engine

arobins commented on lib/org/chromium/webidl/CanonicalizerRunner.js:
```diff
@@ -0,0 +1,53 @@
+      }
+
+      // Insert all of the given ASTs.
+      asts.forEach(function(ast) {
```
Will the message received by this runner contain all of the fragments?

m-cheung replied:
Not anymore. I have rewired the pipelines so each pipeline has their own instance of CanonicalizerRunner. Thus, all messages passing through will belong to the same source.

arobins replied:
Right, but for that source, will the message contain all fragments or some subset of them?

m-cheung replied:
At the moment, each message is a subset of the fragments.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+      name: 'waitTime',
+    },
+    {
+      documentation: `A timer that is used to determine whether waitTime
```
Replace "timer" with "a timeout id returned by setTimeout()". Such an ID should be a positive integer [1]. Let's use `class: 'Int'`.

[1] https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout

m-cheung replied:
In Node, setTimeout doesn't return an int, it returns a `Timeout` object.
https://nodejs.org/dist/latest-v6.x/docs/api/timers.html#timers_class_timeout

I am going to leave this field untouched, but I have updated the description.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+      name: 'timer_',
+    },
+    {
+      documentation: `A map containing canonical IDL files. The key of the map
```
Nit: Use phrasing such as "canonical IDLFile objects". When I first read this, I was thinking "What makes .idl file canonical"? Then I realized you were referring to a class of object.

m-cheung replied:
This has been changed to "canonical AST of definitions" since these aren't actually IDLFile objects.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+  ],
+
+  methods: [
+    {
```
nit: Move private methods to bottom of methods array.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+      documentation: 'Takes two ASTs and merges the members from src to dst.',
+      name: 'merge_',
+      code: function(dst, src) {
+        var isDstPartial = dst.definition.isPartial;
```
Has the type of dst and src already been asserted?

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+          this.error(`Canonicalizer: Two non-partial fragments!
+              In Map: ${foam.json.Pretty.stringify(dst.sources)},
+              Current: ${foam.json.Pretty.stringify(src.sources)}`);
+          throw new Error(`Canonicalizer: Two non-partial fragments were ` +
```
nit: Don't bother mixing string concatenation and template string. It's fine to have extra whitespace in the error string.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+            .concat(src.definition.members);
+
+        // Copy over inheritance information.
+        if (!dst.definition.inheritsFrom)
```
I think this is fine, but wanted to double-check: Only non-partials can have inheritsFrom, correct? (I.e., we don't need a second assertion here to avoid overwriting inheritance).

m-cheung replied:
Yes that is correct. I looked at the grammar for WebIDL and it does not seem like non-partials are allowed to inherit

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+      code: function(ast) {
+        // Find any existing definition with same id (name).
+        var clone = ast.clone();
+        foam.assert(ast.id);
```
Please add an error message as second parameter to `foam.assert()`.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+        } else if (this.Enum.isInstance(ast.definition) ||
+            this.Typedef.isInstance(ast.definition)) {
+          // An Enum or Typedef with the same name was previously defined!
+          console.warn(`Canonicalizer: Encountered two definitions of type` +
```
Ditto: No need to mix string styles.

mdittmer commented on lib/org/chromium/webidl/Canonicalizer.js:
```diff
@@ -0,0 +1,139 @@
+      name: 'timer_',
+    },
+    {
+      documentation: `A map containing canonical IDL files. The key of the map
```
Please describe type more thoroughly. From your code it looks like it's a map: `{ String : Definition | Enum | Typedef | [ Enum | Typedef ] | [ Enum | Typedef ] }`. At the very least, we should document this. In the future we may want to revisit it to simplify as well.

mdittmer commented on lib/org/chromium/webidl/PipelineMessage.js:
```diff
@@ -33,5 +33,9 @@ foam.CLASS({
     },
+    {
+      name: 'canonicalMap',
+      documentation: 'A map of canonical definitions.',
```
Ditto: Describe type inside map.

mdittmer commented on lib/org/chromium/webidl/ast/DefinitionData.js:
```diff
@@ -10,6 +10,20 @@ foam.CLASS({
+        a concrete definition. Otherwise, the AST node represents a canonical
+        definition.`,
+      name: 'isCanonical',
+      value: false,
```
nit: This is the default default. Don't need to explicate it here.

mdittmer commented on lib/org/chromium/webidl/ast/DefinitionData.js:
```diff
@@ -10,6 +10,20 @@ foam.CLASS({
+        non-partial definition. isPartial is set to false if a non-partial
+        definition has been merged in. Otherwise, the value is set to true.`,
       name: 'isPartial',
       value: false,
```
Ditto: default default (for consistency).
